group-generators/helpers/data-providers/rep3/index.ts

```diff
@@ -0,0 +1,97 @@
+import { gql } from "graphql-request";
+import { QueryMembersOutput, QueryMembersInput, IRep3Provider } from "./types";
+
+import { GraphQLProvider } from "@group-generators/helpers/data-providers/graphql";
+import { FetchedData } from "topics/group";
+
+export class Rep3Provider extends GraphQLProvider implements IRep3Provider {
+  constructor() {
+    super({
+      url: "https://api.thegraph.com/subgraphs/name/eth-jashan/rep3-matic",
+    });
+  }
+
+  public async getMembershipHolders({
+    contract,
+    level,
+  }: QueryMembersInput): Promise<FetchedData> {
+    const holders: FetchedData = {};
+
+    if (!this.isValidLevel(level)) {
+      throw new Error(
+        "Invalid level, has to be `all` or a number between 1 and 256"
+      );
+    }
+    const levelSetting =
+      level === "all" ? `level_gte: "1"` : `level: "${level}"`;
+
+    try {
+      const pageSize = 1000;
+      let skip = 0;
+      let continuePaging = true;
+
+      while (continuePaging) {
+        const response: QueryMembersOutput = await this.fetchPage(
+          contract,
+          levelSetting,
+          pageSize,
+          skip
+        );
+
+        if (response.membershipNFTs.length > 0) {
+          response.membershipNFTs.forEach((holder) => {
+            holders[holder.claimer] = 1;
+          });
+          skip += pageSize;
+        } else {
+          continuePaging = false;
+        }
+      }
+    } catch (error) {
+      throw new Error("Error fetching membership holders");
+    }
+    return holders;
+  }
+
+  public async fetchPage(
+    contract: string,
+    levelSetting: string,
+    pageSize: number,
+    skip: number
+  ): Promise<QueryMembersOutput> {
+    const query = gql`{
+      membershipNFTs(
+        where: { contractAddress: "${contract}", ${levelSetting} }
+        first: ${pageSize}
+        skip: ${skip}
+      ) {
+        claimer
+        level
+        tokenID
+      }
+    }`;
+
+    return this.query<QueryMembersOutput>(query, {
+      variables: {
+        contract,
+        levelSetting,
+        pageSize,
+        skip,
+      },
+    });
+  }
+
+  public async getMembershipHoldersCount({
+    contract,
+    level,
+  }: QueryMembersInput): Promise<number> {
+    const holders = await this.getMembershipHolders({ contract, level });
+    return Object.keys(holders).length;
+  }
+
+  private isValidLevel(level: string): boolean {
```

Comment:
This regex checks if the user put any number between 0 and 256 (the allowed levels) or if they used the word "all".
This way, the user can create badges for any level e.g. 
- create badge for level 3 or level 200
- create badge for level 0 (maybe create a badge for inactive users?, probably not?)
- use the word "all" and create a badge for all members except for level 0

Let me know if you prefer this differently

Reply:
Oh no that's perfect.
I though level 0 wasn't accepted because you said at the line 22: "Invalid level, has to be `all` or a number between 1 and 256" (so maybe you need to change this sentence for "between 0 and 256"?)

Reply:
Thank you, that was a mistake. Level 0 is still a level.